message/typeset.go

```diff
@@ -30,6 +30,65 @@ func TypesOf(messages ...dogma.Message) TypeSet {
 	return s
 }
 
+// IntersectionT returns the intersection of the given collections.
+//
+// That is, it returns a set containing only those types that are present in all
+// of the given collections.
+func IntersectionT(collections ...TypeCollection) TypeSet {
+	r := TypeSet{}
+
+	if len(collections) > 0 {
+		collections[0].Each(func(t Type) bool {
+			for _, c := range collections[1:] {
+				if !c.Has(t) {
+					return true
+				}
+			}
+
+			r.Add(t)
+
+			return true
+		})
+	}
+
+	return r
+}
+
+// UnionT returns the union of the given collections.
+//
+// That is, it returns a set containing all of the types present in all of the
+// given collections.
+func UnionT(collections ...TypeCollection) TypeSet {
+	r := TypeSet{}
+
+	for _, c := range collections {
+		c.Each(func(t Type) bool {
+			r.Add(t)
+			return true
+		})
+	}
+
+	return r
+}
+
+// DiffT returns the difference between a and b.
+//
+// That is, it returns a set containing all of the types present in a that are
+// not present in b.
```

Comment:
Should it mention that if `b` does not have any differences to `a`, then an empty set is returned?

Reply:
Again it's not a special case -- that's also true if `b` is a proper superset of `a`.  Maybe I should link to some external docs on the various set operations?

I am re-reading these docs though, and I think I should at least explain what happens when you don't pass any collections at all.